risc0/zkvm/src/prove/mod.rs

```diff
@@ -190,6 +191,70 @@ pub trait Prover {
     fn get_name(&self) -> String;
 }
 
+/// An implementation of a [Prover] that runs proof workloads remotely.
+pub struct RemoteProver {
+    name: String,
+}
+
+impl RemoteProver {
+    /// construct a remote prover. Unlike the [LocalProver], the hal is taken
+    /// care of by the remote prover.
+    pub fn new(name: &str) -> Self {
+        Self {
+            name: name.to_string(),
+        }
+    }
+}
+
+impl Prover for RemoteProver {
+    fn get_name(&self) -> String {
+        self.name.clone()
+    }
+
+    fn get_peak_memory_usage(&self) -> usize {
+        0
+    }
+
+    fn prove_session(&self, _ctx: &VerifierContext, session: &Session) -> Result<SessionReceipt> {
+        log::debug!("prove_session: {}", self.name);
+        let client = bonsai_sdk::Client::from_env()?;
+
+        let session = match &session.bonsai_session_id {
+            Some(id) => bonsai_sdk::SessionId::new(id.clone()),
+            None => return Err(anyhow!("Session does not have a bonsai session ID")),
+        };
+
+        loop {
+            // The session has already been started in the executor. Poll bonsai to see if
+            // the proof request succeeded.
+            let res = session.status(&client)?;
+            if res.status == "RUNNING" {
+                std::thread::sleep(Duration::from_secs(15));
+                continue;
+            }
+            if res.status == "SUCCEEDED" {
+                // Download the receipt, containing the output
+                let receipt_url = match res.receipt_url {
+                    Some(url) => url,
+                    None => return Err(anyhow!("API error, missing receipt on completed session")),
+                };
+
+                let receipt_buf = client.download(&receipt_url)?;
+                let receipt: SessionReceipt = bincode::deserialize(&receipt_buf)?;
+                return Ok(receipt);
+            } else {
+                return Err(anyhow!("Bonsai prover workflow exited: {}", res.status));
+            }
+        }
+    }
+
+    fn prove_segment(&self, _ctx: &VerifierContext, _segment: &Segment) -> Result<SegmentReceipt> {
+        Err(anyhow!(
```

Comment:
Should we consider using https://doc.rust-lang.org/std/macro.unimplemented.html instead of returning a failable error? Im not sure either.

Reply:
yeah... I'm not sure about this. I saw all of the cases that use prove_segment. the unimplemented will panic which I know we're trying to reduce...

Reply:
@flaub do you have any thoughts on this?

Reply:
I think in this case a panic is probably fine since it indicates a programmer mistake. One way to think about it is if this panics, will it bring down a decentralized node? I'd say in this case perhaps, but I'd argue that was a logical mistake by the programmer of the node software itself, and not related to any outside user input.

Reply:
Thanks for your clarification!